src/lib/signature.ts

```diff
@@ -192,12 +198,23 @@ class Signature extends CircuitValue {
   static create(privKey: PrivateKey, msg: Field[]): Signature {
     const publicKey = PublicKey.fromPrivateKey(privKey).toGroup();
     const d = privKey.s;
-    const kPrime = Scalar.random();
+    const kPrime = Scalar.fromBigInt(
+      deriveNonce(
+        { fields: msg.map((f) => f.toBigInt()) },
+        { x: publicKey.x.toBigInt(), y: publicKey.y.toBigInt() },
+        BigInt(d.toJSON()),
+        'testnet'
+      )
+    );
     let { x: r, y: ry } = Group.generator.scale(kPrime);
     const k = ry.toBits()[0].toBoolean() ? kPrime.neg() : kPrime;
-    const e = Scalar.fromBits(
-      Poseidon.hash(msg.concat([publicKey.x, publicKey.y, r])).toBits()
+    let h = hashWithPrefix(
+      prefixes.signatureTestnet,
+      msg.concat([publicKey.x, publicKey.y, r])
     );
+    // TODO: Scalar.fromBits interprets the input as a "shifted scalar"
+    // therefore we have to unshift e before using it
+    let e = unshift(Scalar.fromBits(h.toBits()));
```

Comment:
So this seems like a bug in `Scalar.fromBits`, which is somewhat hard to resolve in that function directly, since this function would need to unshift the scalar in a SNARK-compatible way. It seemed easier to unshift out of the snark (here) and do the shifting at the level of curve points when we're in the snark (see `verify` below)